### ISSUE TITLE
Charts rendered in hidden DIVs will not display until browser resize

### ISSUE THREAD

**Issue by codyseibert** [jtblin/angular-chart.js#29]
If there is already a work around for this, please let me know, but here is my use case / issue.

I am using foundation's tabset directive to build up tabs and their corresponding containers.  I have some the chart.js graphs being defined inside a hidden container; therefore, the charts have no container width or height when it comes to rendering the graphs.  

Would there be a way to achieve this?  Could you potentially add in a "reset" watch to the angular-chart.js which will destroy and recreate the chart?

**Comment by jtblin** [comment]
This looks similar to #17 which itself is same as https://github.com/nnnick/Chart.js/issues/512. 

Could you just set the data when the tab is shown/active? That will draw the chart when the tab is first shown.

**Comment by leonetnl** [comment]
Check out my fork (dont know how to link to it, my first contribution on github). You can now send a angular broadcast '$reload' to repaint the graphs.

**Comment by jtblin** [comment]
Cool, why not submitting a PR? ;)

https://help.github.com/articles/using-pull-requests/

Links to a PR is as simple as putting a link to it in the comment.

**Comment by jaman1020** [comment]
I've added @leonetnl 's changes to my personal copy of angularjs, and the broadcast works, but it doesn't seem to redraw the chart. I know this is more of a Chart.js issue, but I've been pulling my hair out with it for a few hours now and can't get it working. 

If the original chart is drawn successfully, I can call chart update/create/whatever. However, if the initial draw call comes when the tab is *not* active, it doesn't draw. I get this error - 
    Uncaught IndexSizeError: Failed to execute 'arc' on 'CanvasRenderingContext2D': The radius provided (-0.5) is negative.

No matter how I try to update the chart - whether I update it, destroy it, create a new one, it just doesn't work. Its almost like the canvas permanently fails the draw, and refuses to reload unless part or all of the html itself is reloaded.

I'm willing to integrate the changes and submit a pull request myself, but I can't get it drawing correctly.

**Comment by jaman1020** [comment]
So I've figured out the cause - although not a solution (at least for me). It looks like the elem variable in the directive is never updated when the element changes height, so it retains the 0px height it gets when the div is not visible. I'm personally using angular-ui bootstrap tabs, so when I tab back to the original tab, the elem variable has a 0px height, although the actually element it is referring to no longer does. That is why the destroy/create method recommended for chartjs doesn't work, even if its called when the div is visible.

**Comment by shak18** [comment]
my only solution so far to this has been to use a directive with the $compile injector and delete and repaint the canvas element in the dom when it becomes visible maybe you guys can try a similar work arround

**Comment by bornkiller** [comment]
The same problem encountered, it seems that this issue not resolved yet.

Personally, I transfer the content with chart (which should place between `tab`) into another standalone  `<div></div>`  container, and use `ng-show` and tab `select()`, `deselect()` to control the visibility. 

```html
<div class="container" style="width: 60%;" ng-controller="MainCtrl">
   <tabset justified="true">
       <tab heading="chart" select="select()" deselect="deselect()">
       </tab>
       <tab heading="info">
           The beautiful movie "Whiplash"!
       </tab>
   </tabset>
   <div ng-show="chartTabShow">
       <canvas id="line" class="chart chart-line" data="data" labels="labels" series="series">
       </canvas>
    </div>
</div>
```

```javascript
$scope.select = function() {
   $scope.chartTabShow = true;
};
$scope.deselect = function() {
   $scope.chartTabShow = false;
};
```

To be honest, it's not good solution, even for me. I consider this way as temporary solution for urgent situation, not recommended.

**Comment by jtblin** [comment]
Have you tried just initialising the data when you select the tab?

```
$scope.select = function() {
   $scope.data = savedData;
};
```

**Comment by IgorUzhviev** [comment]
jtblin, I tried that - unfortunately it did not work as well.

**Comment by germanger** [comment]
This problem reminds me to an "ng-grid" inside a "bootstrap-ui tab".
I typically solve that problem with an uggly `$(window).resize()` when the tab is selected...

eg:

        $scope.ngGridFIx = function() {
               $(window).resize()
        }

       <tab select="ngGridFix()"> ... </tab>

BUT HERE......  the `$(window).resize()` is not working for me!... and it is weird because if I manually (not programatically) resize the window, the chart is fixed.

So,... what can I do?

**Comment by jtblin** [comment]
Here is a jsbin for @bornkiller's solution: http://jsbin.com/sopupij/1/edit?html,js,output. Not ideal but it works.

**Comment by jtblin** [comment]
And here is an example by delaying setting the data until the tab is selected + wrapped in a `$timeout`: http://jsbin.com/sopupij/2/edit. Not ideal either but...

**Comment by germanger** [comment]
Thanks @jtblin 
The $timeout solution worked well when the chart is in a bootstrap-ui modal, but I had to add 100ms as the delay parameter (if not, the result was random: it worked sometimes)

**Comment by karthikeyanthangavel** [comment]
Hi  jtblin,

I'm using angular-chart.js in our application. 

Question: Instead of tooltip i want to show label values in bar/line/pie charts .Please give me a solution as soon as possible ?

**Comment by mikerennick** [comment]
Several have mentioned the solution of setting the data after the tab/accordion/previously-hidden-div is brought into focus.  This works for me, but only after setting a timeout (in my case 1 sec).  Here are the main things to keep in mind:

* since you are using bootstrap tabs or accordion, and each panel has it's own scope, you cannot simply place the data for the chart directly onto your $scope in the controller.  I set up my data like this:
```
    $scope.chartSettings = {
                labels: ["January", "February", "March", "April", "May", "June", "July"],
                series: ['Series A', 'Series B'],
                onClick: function (points, evt) {
                    console.log(points, evt);
                }
            }

    the HTML....

    <canvas id="line" class="chart chart-line" data="chartSettings.data" labels="chartSettings.labels" legend="true" series="chartSettings.series" options="chartSettings.options"  click="chartSettings.onClick"></canvas>
```
* Since the directive for the chart watches various pieces of data and resets the chart when that data changes, you can set the data element for the chart after the tab/accordion comes into focus...like this
```
    //a snippet of the HTML for the accordion...

     <accordion-group is-open="false">
        <accordion-heading>
            <div ng-click="opened(1)">
                Chart <i class="pull-right glyphicon" ng-class="{'glyphicon-chevron-down': groups[1].open, 'glyphicon-chevron-right': !groups[1].open}"></i>
            </div>
    .... etc....

    $scope.opened = function (groupIndex) {
                $scope.groups[groupIndex].open = !$scope.groups[groupIndex].open;

                if (groupIndex === 1) { //this is the accordion that holds the chart

                    $timeout(updateChart, 1000);
                }

                return true;
            }

    var updateChart = function() {
                $scope.chartSettings.data = [
                        [65, 59, 80, 81, 56, 55, 40],
                        [28, 48, 40, 19, 86, 27, 90]
                ];
            }
```
The timeout was necessary as we need to wait until the accordion or the tab has a true height...1 second seemed to work for me.

**Comment by themsaid** [comment]
@germanger 

You can use this instead

```
$scope.ngGridFIx = function() {
    window.dispatchEvent(new Event('resize'));
}
```

It worked for me

**Comment by dudewheresmycode** [comment]
Any update on this? I'm also having trouble trying to draw a chart in a "ng-hide" hidden situation to no avail.

**Comment by jtblin** [comment]
I don't think that can be easily fixed unfortunately otherwise that would be already done. If the workaround described above doesn't work for you, another option is to set responsive to false in the options and set a height and width directly on the canvas.

**Comment by rfink** [comment]
+1

**Comment by gregor-srdic** [comment]
@jtblin Your workaround in the example does not work if you select "info" tab, then resize the display area and then switch to "chart" tab. The graph is not displayed. In my ionic app, I think I solved this issue with 
`$timeout(function(){
 window.dispatchEvent(new Event('resize'));
});`
when the chart tab becomes active

**Comment by Jonesb** [comment]
I had to update the createChart function in the angular-chart.js which stopped the ChartJs.Chart() function from zeroing the size of the canvas when it wasn't visible.

This is what I inserted
```javascript
if (ctx.canvas['offsetHeight'] == 0)
          return;
```

This is the full function

```javascript
     function createChart (type, scope, elem) {

      if (! scope.data || ! scope.data.length) return;
      scope.getColour = typeof scope.getColour === 'function' ? scope.getColour : getRandomColour;
      scope.colours = getColours(type, scope);
      var cvs = elem[0], ctx = cvs.getContext('2d');

      if (ctx.canvas['offsetHeight'] == 0)
          return;

      var data = Array.isArray(scope.data[0]) ?
        getDataSets(scope.labels, scope.data, scope.series || [], scope.colours) :
        getData(scope.labels, scope.data, scope.colours);
      var options = angular.extend({}, ChartJs.getOptions(type), scope.options);
      var chart = new ChartJs.Chart(ctx)[type](data, options);
      scope.$emit('create', chart);

      ['hover', 'click'].forEach(function (action) {
        if (scope[action]) cvs[action === 'click' ? 'onclick' : 'onmousemove'] = getEventHandler(scope, chart, action);
      });
      if (scope.legend && scope.legend !== 'false') setLegend(elem, chart);
      return chart;
    }
```

**Comment by Jonesb** [comment]
I also had to add a select() function on the tab with set a $timeout function to apply the data

```javascript

$scope.selectFunction = function () {

            $timeout(function () {

                $scope.data = [$scope.Value1, $scope.Value2, $scope.Value3];

            }, 200);
            
        }
```

**Comment by Jonesb** [comment]
After further usage the above solution doesn't work with resize events when the chart is hidden as the canvas gets zeroed.
I think this is a general problem with Chart.js not angular chart and should be fixed in that library.  
Rolling back the CreateCanvas function and Setting responsive = false fixes the issue.
Sorry I'm new to the library:)

**Comment by SandrineProusteau** [comment]
I have several graphs in several foundation-angular-tabs. I found a workaround like this :
In angular-chart.js in "link" function, add watcher on tabs ($scope.tabs of foundationangular tabs demo) which indicate the active tab 
```javascript
scope.$parent.$watch('tabs', function (newVal, oldVal) {
            setTimeout(function(){
              var chartType = type || scope.chartType;
              if (! chartType) return;
              if (chart) {
                if (canUpdateChart(newVal, oldVal)) return updateChart(chart, newVal, scope);
                chart.destroy();
              }
              chart = createChart(chartType, scope, elem);
            },1000);
          }, true);
```
in this watcher, add a setTimout in order that the DOM is updte for check the offsetHeight of the canvas!

**Comment by CarbonFactory** [comment]
Setting the option `responsive` to `false` did not solve the issue. Charts in inactive tabs do not get displayed when made active.

**Comment by jtblin** [comment]
@CarbonFactory the chart not appearing when in a hidden container e.g. non active tab is due to the `responsive` mode as in this case, chart.js tries to calculate the dimensions based on the container dimensions, see issue in chart.js and proposed workaround by the author of the lib: https://github.com/nnnick/Chart.js/issues/512#issuecomment-52506918 (which is what I recommended i.e. initialising the chart on tab show).

If `responsive` doesn't work for you, there may be another issue with your setup maybe?

**Comment by nsaje** [comment]
Setting `responsive` to false didn't work for me either, but the timeout workaround did.

**Comment by jtblin** [comment]
@nsaje thanks for the report, would you have a quick [repro case](http://jsbin.com/dufibi/3/edit?html,js,output) for the tab (?) responsive issue?

**Comment by darrynten** [comment]
This issue has been driving me crazy all day. At first I thought I was doing something wrong, but I'm glad to hear I'm not the only one.

Does anyone have any other workarounds? The above ones don't work for me.

**Comment by FelixCoutinho** [comment]
@darrynten I solve with this solution...

http://pastebin.com/wZMz8DnJ

**Comment by juanchavezlive** [comment]
The solution with the "$timeout" works. However it's a dirty trick.

**Comment by benbrownlol** [comment]
Yeah the $timeout fix is definitely the best way to do it if anyone else comes across this issue.

```
// Referral Sources - graph fix applied
$scope.openReferralSources = function() {
	$mdSidenav('referralSources').open();
	$timeout(function() {
		$scope.graphs.testGraph = {
			labels: ["Word of Mouth", "Dentistry Magazine", "Email Marketing", "Radio Advert", "TV Advert", "GP Referral"],
			data: [[10, 4 , 7, 9, 11, 3]],
			colours: ["#0095D2"]
		}
	});
};
```

**Comment by vchuquimia** [comment]
The $(window).resize() solution  it seems is not working in my case,  what I'm trying to do is to show a Graph and its Data in a Collapse/Expand Viewer. by clicking on 'right' the graph is going to be maximized and by clicking on 'left' the data table.

this is the jsBin for this issue : [http://jsbin.com/qopife/edit?js,output](http://jsbin.com/qopife/edit?js,output)


what I have tried : chart.resize(chart.render) and $(window).resize() 
Any ideas ?

**Comment by giannoug** [comment]
It seems that I'm also affected by this bug. Using plain old ng-show for some tabs and the graphs are not showing. Setting `responsive` to `true` fixes the problem, but you lose the responsive features. I had no luck with the timeouts.

**Comment by h-xyl** [comment]
I used ng-if instead of ng-class or ng-show/hide. It worked.

**Comment by gregor-srdic** [comment]
I use the following code in a $timeout to fire a window resize event when a tab with chart becomes active:

```javascript 
function emitWindowResizeEvent(){
    var eventName='resize', el=window, event;
    if(document.createEvent){
        event = document.createEvent('HTMLEvents');
        event.initEvent(eventName,true,true);
    }else if(document.createEventObject){// IE < 9
        event = document.createEventObject();
        event.eventType = eventName;
    }
    event.eventName = eventName;
    if(el.dispatchEvent){
        el.dispatchEvent(event);
    }else if(el.fireEvent && htmlEvents['on'+eventName]){// IE < 9
        el.fireEvent('on'+event.eventType,event);// can trigger only real event (e.g. 'click')
    }else if(el[eventName]){
        el[eventName]();
    }else if(el['on'+eventName]){
        el['on'+eventName]();
    }
} 
```

**Comment by jtblin** [comment]
For the record, the issue upstream is https://github.com/nnnick/Chart.js/issues/961

**Comment by jtblin** [comment]
So I pushed a fix to master which should remove the issue with responsive charts not being created when placed in a hidden div, tab, etc. I added a working example in `examples/tabs.html`. 

Would someone who encountered the issue be willing to give it a try with the latest assets from master before I publish a new version?

**Comment by juanchavezlive** [comment]
@jtblin I'm your fan :) :)

**Comment by jtblin** [comment]
Thank you! :flushed:  Just make sure it works for you before you say that though :D

**Comment by emakarov** [comment]
Read the thread and didn't find an answer. 
So posting here My workaround - with jQuery plugin that tracks "inview" event -

$("#barChart").on('inview', function () {                   
    if (typeof myNewChart == "undefined" ){
      ctx = document.getElementById("barChart").getContext("2d");
      myNewChart = new Chart(ctx).Bar(barData, barOptions);
    } else {
      myNewChart.destroy;
      myNewChart = new Chart(ctx).Bar(barData, barOptions);
    }
});

**Comment by emakarov** [comment]
Read the thread and didn't find an answer.
So my workaround - 

var ctx;
var myNewChart = undefined;

$("#barChart").on('inview', function () {                   
    if (typeof myNewChart == "undefined" ){
      ctx = document.getElementById("barChart").getContext("2d");
      myNewChart = new Chart(ctx).Bar(barData, barOptions);
    } else {
      myNewChart.destroy();
      myNewChart = new Chart(ctx).Bar(barData, barOptions);
    }
});

**Comment by atmosuwiryo** [comment]
Using @leonetnl solution work best for me.
Broadcast $reload on tab/accordion click.

**Comment by lobobobo** [comment]
THANKS @jtblin $timeout worked in modal for chart line

**Comment by jtblin** [comment]
@lobobobo this shouldn't have been necessary anymore. Did you face an issue with a recent version?

**Comment by VoidVolker** [comment]
Same issue. In my case I have hidden layer (display: none) with chart and want to draw it in background. After activating this layer I have canvas with 0 size. I checked the chart object and found there 2 NaN:

    aspectRatio: NaN
    height: NaN

Looks like there somewhere is not type safe code.

**Comment by VoidVolker** [comment]
And NaN is generates is here:

    document.defaultView.getComputedStyle(element).getPropertyValue(dimension);
    ...
    var width = this.width = context.canvas.width;
    var height = this.height = context.canvas.height;
    this.aspectRatio = this.width / this.height; // 0 / 0 = NaN

**Comment by VoidVolker** [comment]
Dropped chart.js, because I found: http://c3js.org/gettingstarted.html - it larger, but it work out of the box with hidden layers.

**Comment by lobobobo** [comment]
i am using chart.js with ionic, when i show and hide the graph the graph shows for the first time but when i hide it and shows again the graph disappears....
Please guide me !

**Comment by lobobobo** [comment]
In chartjs destroy() clear() removeData() nothing is working why ?

**Comment by josmanperez** [comment]
Hello! I've been dealing with this for two days long.. finally I've set the responsive flag to false and the Line Chart shows at the beginning when the modal window is shown.. but I really want to be able to use the responsive feature but with this flag set to true I have to resize the browser after showing the modal window to see the graph...

I read all this document but cannot find a solution that fits because when responsive set to false it works.

thank you so much!

**Comment by atmosuwiryo** [comment]
Have you tried to broadcast $reload on tab/accordion click?
It works on my.

**Comment by harishgadiya** [comment]
hello i think i have fix this just make your chart in a function and invoke after the function in your controller

**Comment by T3chn0Viking** [comment]
another simple fix is disabling the caching of the specific page that contains the chart, I did this by disabling the caching in the state provider.

$stateProvider.state('myState', {
   cache: false,
   url : '/myUrl',
   templateUrl : 'my-template.html'
})

**Comment by Dmytraw** [comment]
@themsaid man, u saved my arse!:)

**Comment by victoralfonsoperez** [comment]
resizing the container with a simple css animation worked for me

**Comment by xdega** [comment]
@victoralfonsoperez Could you share the CSS workaround that you used?

**Comment by johncrumpton** [comment]
I added this when the link was clicked for showing the bar chart:
`$("#bar-chart").css("height","721px");`

It's 1px larger than the existing bar-chart box

**Comment by rameshkumarv** [comment]
**Solved the issue by using ng-if instead of ng-show/ng-hide.**
In my case, I used ng-show to show the loader until the HTTP request to complete, then used ng-show to show the graph content, this collapsed the bar chart .

Tried with timeouts worked 90% of the time, but failed  some times.
Disabled the cache, worked fine when switching between states/pages but failed when the page loads initially.

**Comment by SANTHK** [comment]
The solution given by @h-xyl works well for me. Thanks!

**Comment by FredH76** [comment]
I still had the same problem : when coming back from a side menu, my graph blinked and disapeared.
I implemented a solutio that worked well : I put the graph initialization in a  timeout:
$document.ready(function() {
    $timeout(function() {
        var ctx = document.getElementById("weightChart");
        vm.weightChart = new Chart(ctx, weightConfig);
      },
      50
    );
  });

**Comment by rakeshkumartiwari** [comment]
If chartjs is not showing data that means that div is hidden where you put your canvas tag and one more reason which I found that I didn't put my whole methd in ajax succecss method thats why I am getting that error when I put my whole chartjs related thing in ajax success methd, it was fixed.

**Comment by marianomoretti** [comment]
I had this problem using angular and bootstrap. I remembered that Chart.js calculates the size chart basing on the parent div when you are using the responsive: true option. So, I set height to the parent div and that makes that Chart.js can calculate the size chart. The result is that the chart is drew even when is hidden.

**Comment by tomEwen** [comment]
for me adding:
setTimeout(chart.update, 10);
after the creation of the chart worked.

**Comment by wbecker64** [comment]
I had the same issue with a nvd3 graph inside a uib-tab not displayed at first.
I've added this in the controller : 

        // Listen to the changes of visible attribute of the chart.
        $scope.$watch(function () {
            return $element.is(':visible');
        }, function (newValue) {
            if (chartRef && newValue) {
                // Update the chart when it becomes visible (newValue is true)
                chartRef.update();
            }
        });

**Comment by piilupart88** [comment]
I had a similar problem rendering an angularjs directive with chartjs on a popup modal. Rendering a modal took some time, therefore the container of the chart had no parameters at the time it was calculated.
Tried many things for a couple a days and finally found a solution that worked for me.

My solution was to **increase the priority **of the** directive** with the chart so that it would be rendered after the modal it was on.
